src/builtin.c

```diff
@@ -1446,65 +1446,108 @@ static jv f_strptime(jq_state *jq, jv a, jv b) {
   return r;
 }
 
-#define TO_TM_FIELD(t, j, i)                    \
-    do {                                        \
-      jv n = jv_array_get(jv_copy(j), (i));     \
-      if (jv_get_kind(n) != (JV_KIND_NUMBER)) { \
-        jv_free(n);                             \
-        jv_free(j);                             \
-        return 0;                               \
-      }                                         \
-      t = jv_number_value(n);                   \
-      jv_free(n);                               \
-    } while (0)
+/*
+ * Utility function for jv2tm() that returns `ifnan` if `v` is not a number in
+ * any way, and clamps the value range of `v` to `imin` and `imax`.
+ */
+static int toint(jv v, int imin, int imax, int ifnan) {
+  if (jv_get_kind(v) != (JV_KIND_NUMBER) || jvp_number_is_nan(v)) {
+    jv_free(v);
+    return ifnan;
+  }
+  double n = jv_number_value(v);
+  jv_free(v);
+  if (isinf(n)) {
+    if (n < 0.0)
+      return imin;
+    return imax;
+  }
+  if (n < imin)
+    return imin;
+  if (n > imax)
+    return imax;
+  return (int)n;
+}
 
 static int jv2tm(jv a, struct tm *tm, double *frac, char **freeme) {
   int ret = 1;
 
   *freeme = NULL;
   memset(tm, 0, sizeof(*tm));
-  TO_TM_FIELD(tm->tm_year, a, 0);
+  if (   (tm->tm_year = toint(jv_array_get(jv_copy(a), 0),
```

Comment:
ISO 8601 has an option for 5-digit years, but we'll go up to just 4, which means up to `9999`.  For now we won't support time values before 1970 (though we should).